python-frontend/src/main/java/org/sonar/python/api/IPythonGrammarBuilder.java

```diff
@@ -40,20 +45,32 @@
 import static org.sonar.python.api.PythonGrammar.RAISE_STMT;
 import static org.sonar.python.api.PythonGrammar.RETURN_STMT;
 import static org.sonar.python.api.PythonGrammar.SIMPLE_STMT;
+import static org.sonar.python.api.PythonGrammar.STATEMENT;
 import static org.sonar.python.api.PythonGrammar.TESTLIST_STAR_EXPR;
 import static org.sonar.python.api.PythonGrammar.YIELD_EXPR;
 import static org.sonar.python.api.PythonGrammar.YIELD_STMT;
+import static org.sonar.python.api.PythonTokenType.IPYNB_CELL_DELIMITER;
 import static org.sonar.python.api.PythonTokenType.NEWLINE;
 
 public class IPythonGrammarBuilder extends PythonGrammarBuilder {
 
   @Override
-  protected void setupRules(LexerfulGrammarBuilder b) {
-    lineMagic(b);
-    super.setupRules(b);
+  public Grammar create() {
+    LexerfulGrammarBuilder b = LexerfulGrammarBuilder.create();
+
+    b.rule(FILE_INPUT).is(b.zeroOrMore(b.firstOf(CELL, MAGIC_CELL, NEWLINE, IPYNB_CELL_DELIMITER)), EOF);
+
+    setupIPythonNotebooksRules(b);
+    setupCommonRules(b);
+
+    b.setRootRule(FILE_INPUT);
+    return b.buildWithMemoizationOfMatchesForAllRules();
```

Comment:
probably it is better to extend `setupFileInput` method instead of override whole `create` method with all the code almost the same as in the parent builder

Reply:
As discussed, I went with a slightly different approach and just decided to define all rules in a `setupRules` method and to overload that. I also created a `fileInput` method to overload as well.